pyanaconda/modules/storage/bootloader/systemd.py

```diff
@@ -0,0 +1,190 @@
+
+
+    def write_device_map(self):
+        """Write out a device map containing all supported devices."""
+        log.info("systemd.py: write_device_map")
+        map_path = os.path.normpath(conf.target.system_root + self.device_map_file)
+        if os.access(map_path, os.R_OK):
+            os.rename(map_path, map_path + ".anacbak")
+
+        devices = self.disks
+        if self.stage1_device not in devices:
+            devices.append(self.stage1_device)
+
+        for disk in self.stage2_device.disks:
+            if disk not in devices:
+                devices.append(disk)
+
+        devices = [d for d in devices if d.is_disk]
+
+        if len(devices) == 0:
+            return
+
+    # copy console update from grub2.py
+    def write_config_console(self, config):
+        log.info("systemd.py: write_config_console")
+        if not self.console:
+            return
+
+        console_arg = "console=%s" % self.console
+        if self.console_options:
+            console_arg += ",%s" % self.console_options
+        self.boot_args.add(console_arg)
+
+
+
+    def write_config(self):
+        log.info("systemd.py: write_config systemd start")
+        # So at this point, all this is pretty much redundant
+        self.write_config_console(None)
+
+        # just rewrite the loader.conf
+        #cfg_file = config_file(self)
+        config_path = "%s%s" % (conf.target.system_root, "/boot/efi/loader/loader.conf")
+        log.info("systemd.py: write_config systemd loader conf : %s ", config_path)
+
+        # outside of resetting the default, this file should probably be moved to
+        # stubby as well?
+        config = open(config_path, "w")
+        config.write("timeout 3\n")
+        # create a default entry? No need right now..
+        # config.write("default 4142287a09f64398a50ac387bf9e1fca-*\n")
+        config.write("#console-mode keep\n")
+        config.close()
+
+        # do the bls command line here?
+        # update /etc/kernel/cmdline
+        # should look something like "root=UUID=45b931b7-592a-46dc-9c33-d38d5901ec29 ro  resume=/dev/sda3"
+        config_path = "%s%s" % (conf.target.system_root, "/etc/kernel/cmdline")
+        log.info("systemd.py: write_config systemd commandline : %s ", config_path)
+        config = open(config_path, "w")
+        args = " ".join(list(self.boot_args._arguments))
+        log.info("systemd.py: systemd used boot args: %s ", args)
+        # should run this through blkid to get UUID... but what is going on with the above args...
+        args += " root=/dev/mapper/fedora_fedora-root"
+        config.write(args)
+        config.close()
+        # rather than creating a mess in python lets just
+        # write the options above, and run a script which will merge the
+        # boot cmdline (after stripping inst. and BOOT_) with the anaconda settings
+        rc = util.execWithRedirect(
+            "/usr/sbin/updateloaderentries.sh",
```

Comment:
I see that the `ZIPL` class has an ` update_bls_args()` function that does something similar. Maybe there could be a common helper instead of having a separate script?

Reply:
This is the one piece I looked at for a half day or so, but haven't managed to merge together. I think in general you are right, both are doing something quite similar (updating existing BLS entries) but there are a few subtle differences that will need to be merged around how the command line itself is generated. Worse I suspect both have a fundamental error, in that they aren't filtering for newly installed entries which could be a problem.
I think at this point, this is probably the largest remaining issue.